arch/riscv/lib/string.c

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * arch/riscv/lib/string.c
+ *
+ * Copyright (C) 2021 Matteo Croce
+ *
+ * string functions optimized for 64 bit hardware which doesn't
+ * handle unaligned memory accesses efficiently.
+ *
+ * May be freely distributed as part of Linux.
+ */
+
+#include <linux/types.h>
+#include <linux/module.h>
+
+union types {
+	u8 *u8;
+	u16 *u16;
+	u32 *u32;
+	u64 *u64;
+	unsigned long *ulong;
+	uintptr_t ptr;
+};
+
+union ctypes {
+	const u8 *u8;
+	const u16 *u16;
+	const u32 *u32;
+	const u64 *u64;
+	unsigned long *ulong;
+	uintptr_t ptr;
+};
+
+/**
+ * memcpy - Copy one area of memory to another
+ * @dest: Where to copy to
+ * @src: Where to copy from
+ * @count: The size of the area.
+ *
+ * You should not use this function to access IO space, use memcpy_toio()
+ * or memcpy_fromio() instead.
+ */
+void *memcpy(void *dest, const void *src, size_t count)
+{
+	static const void *labels[] = {
+		&&u64, &&u8, &&u16, &&u8,
+		&&u32, &&u8, &&u16, &&u8,
+	};
+	union types d = { .u8 = dest };
+	union ctypes s = { .u8 = src };
+#ifdef HAVE_EFFICIENT_UNALIGNED_ACCESS
+	int distance = 0;
+#else
+	const int mask = BITS_PER_LONG / 8 - 1;
+	int distance = (src - dest) & 7;
+
+	for (; count && d.ptr & s.ptr & mask; count--)
+		*d.u8++ = *s.u8++;
+#endif
+
+	goto *labels[distance];
+
+u64:
+#if BITS_PER_LONG == 64
+	for (; count >= 8; count -= 8)
+		*d.u64++ = *s.u64++;
+#endif
+
+u32:
+	for (; count >= 4; count -= 4)
+		*d.u32++ = *s.u32++;
+
+u16:
+	for (; count >= 2; count -= 2)
+		*d.u16++ = *s.u16++;
+
+u8:
+	while (count--)
+		*d.u8++ = *s.u8++;
+
+	return dest;
+}
+EXPORT_SYMBOL(memcpy);
+
+void *__memcpy(void *dest, const void *src, size_t count)
+{
+	return memcpy(dest, src, count);
+}
+EXPORT_SYMBOL(__memcpy);
+
+/**
+ * memmove - Copy one area of memory to another
+ * @dest: Where to copy to
+ * @src: Where to copy from
+ * @count: The size of the area.
+ *
+ * Unlike memcpy(), memmove() copes with overlapping areas.
+ */
+void *memmove(void *dest, const void *src, size_t count)
+{
+	if (dest < src || src + count <= dest)
+		return memcpy(dest, src, count);
+
+	if (dest > src) {
+		const char *s = src + count;
+		char *tmp = dest + count;
+
+		while (count--)
+			*--tmp = *--s;
+	}
+	return dest;
+}
+EXPORT_SYMBOL(memmove);
+
+void *__memmove(void *dest, const void *src, size_t count)
+{
+	return memmove(dest, src, count);
+}
+EXPORT_SYMBOL(__memmove);
+
+/**
+ * memset - Fill a region of memory with the given value
+ * @s: Pointer to the start of the area.
+ * @c: The byte to fill the area with
+ * @count: The size of the area.
+ *
+ * Do not use memset() to access IO space, use memset_io() instead.
+ */
+void *memset(void *s, int c, size_t count)
+{
+	const int bytes_long = BITS_PER_LONG / 8;
+	u8 cc[] = { [0 ... bytes_long-1] = c };
+	union ctypes src = { .u8 = cc };
+	union types dest = { .u8 = s };
+
+#ifndef HAVE_EFFICIENT_UNALIGNED_ACCESS
```

Comment:
Ditto

Reply:
So it turns out that memcpy() was working correctly, while memset() was generating unaligned accesses.
I daw memset on top of `perf top` but I couldn't explain why.
Thanks!

Reply:
Well no, that shouldn't have effect anyway.
I see memset always being called when transmitting small packets, maybe it's normal